### PR TITLE
 detect: add requires keyword - v3

### PR DESCRIPTION
**Pull request by jasonish**
Add a new rule keyword "requires" that allows a rule to require specific
Suricata versions and/or Suricata features to be enabled.

Example:

  requires: feature geoip, version >= 7.0.0, version < 8;

Feature: https://redmine.openinfosecfoundation.org/issues/5972

Previous PR: https://github.com/OISF/suricata/pull/9919

Changes from previous PR:
- Don't allow duplicate SID

A pre-scan/requires checking mode was adding to the parser. When the requires
flag is set, all the *parsing* for syntax is done, but no validation/setup is
done except for the `requires` and `sid` keywords. This allows `requires` to be
specified after keywords that the engine may not now about yet. However, the
rule still needs to be properly formatted when comes to syntax, quoting,
escaping, etc.

Parsing the SID was also added to the pre-scan, so now any keyword parser
(except requires and sid) can have access to the sid for logging purposes.

The the following rules will now be skipped as they have requirements that are
not satisfied by Suricata:
```
alert vxlan any any -> any any (requires: version >= 10; sid:1;)
alert udp any any -> any any (vxlan_vni:10; requires: version >= 10; sid:2;)
alert http any any => any any (requires: version >= 10; sid:3;)
alert tcp any any -> any any (frame:smtp.not_supported; requires: version >= 10; sid:4;)
```

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1505

### CODE REVIEW

File: src/detect-parse.c [hunk 1]
```diff
@@ -1452,12 +1467,7 @@ static int SigParse(DetectEngineCtx *de_ctx, Signature *s,
         SCReturnInt(-1);
     }
 
-    s->sig_str = SCStrdup(sigstr);
-    if (unlikely(s->sig_str == NULL)) {
-        SCReturnInt(-1);
-    }
-
-    int ret = SigParseBasics(de_ctx, s, sigstr, parser, addrs_direction);
+    int ret = SigParseBasics(de_ctx, s, sigstr, parser, addrs_direction, requires);
```
Comment (inashivb) [inline src/detect-parse.c:1471]: Is there no overhead for the parsing of tokens done twice per rule and creating duplicates of the sigstr?

**Review by inashivb — COMMENTED** [review]
Thanks for this. :)
I have one inline and the following questions.
1. Why are we so keen on finding syntactic errors before requires? Will the result not be failure in loading of the rules if an issue was found post the requirements check?
2. Why are we reparsing parts of rule (like action, proto, etc) and doing some semantic checks twice (w scan-only and then proper mode)?

Trying to understand the decision of reusing the code that makes double checks instead of a separate function called like any other rule keyword but before all the other keywords.

File: doc/userguide/rules/meta.rst [hunk 2]
```diff
@@ -211,3 +211,29 @@ The format is::
 If the value is src_ip then the source IP in the generated event (src_ip
 field in JSON) is the target of the attack. If target is set to dest_ip
 then the target is the destination IP in the generated event.
+
+requires
```
Comment (victorjulien) [inline doc/userguide/rules/meta.rst:215]: I think it should be mention that this is parsed first, before any other syntax validation is done

**Review by victorjulien — CHANGES_REQUESTED** [review]
minor doc comment

As mentioned in https://github.com/OISF/suricata-verify/pull/1505, I would like to see more tests for more scenarios

File: doc/userguide/rules/meta.rst (hunk 2 above)
Comment (jasonish) [reply doc/userguide/rules/meta.rst:215]: > When parsing rules, the parser attempts to process the 'requires'
keywords before others. This allows it to occur after keywords that
may only be present in specific versions of Suricata, as specified by
the 'requires' statement. However, the keywords preceding it must
still adhere to the basic known formats of Suricata rules.

File: src/detect-parse.c (hunk 1 above)
Comment (jasonish) [reply src/detect-parse.c:1471]: There is, as it does run through the rules 2x. But its very fast.  On my desktop the loading of 68693 rules took 36 seconds. The total time spent in this first pass was only 0.78 seconds of that.  The actual "parsing" of the rules is very fast, its the `Setup()` functions that take most of the time which is skipped in the first pass, except for `requires` and `sid`. This was with ASAN enabled.

The overhead is more present on a slow Raspberry PI where the prepares takes 1.4 seconds, and the total parse is 9 seconds (no ASAN this time).

**Comment by jasonish** [comment]
> Thanks for this. :) I have one inline and the following questions.
> 
>     1. Why are we so keen on finding syntactic errors before requires? Will the result not be failure in loading of the rules if an issue was found post the requirements check?
> 
>     2. Why are we reparsing parts of rule (like action, proto, etc) and doing some semantic checks twice (w scan-only and then proper mode)?
> 
> 
> Trying to understand the decision of reusing the code that makes double checks instead of a separate function called like any other rule keyword but before all the other keywords.

The only way to find the `requires` statement is to parse our way to it. In order to do that, we have to understand the structure of a rule.

A cleaner approach would be to do that "parse" in one pass to create an abstract syntax tree of sorts. Then scan that for `requies` and `sid`.  Then do a *compilation* phase on the already created abstract syntax tree. But that would be a huge refactor.  As we parse and compile in the same stage right now, this was the simplest approach that may also be worth of a backport.

**Comment by jasonish** [comment]
Continued at https://github.com/OISF/suricata/pull/9977.

**Pull request closed without merging**